Use SQLITE transactions when inserting PCAP data in `recap`.

aloisklink opened the pull request:
Modifies the recap executable to pipe all of the PCAP data into the sqlite database using a single SQLITE transaction.

This has two benefits:

1. On an error, the entire transaction is rolled-back, so we won't have a half imported recap file.
2. Running recap is much, much, much, much faster.
   **A quick local test went from 12.0KiB/s to 40.4 MiB/s, which is a ~3300x increase in speed!**

### Benchmarks

When using the `cmake --preset linux` setting:

#### Before (commit 12a7aa444da23f8225288aafe67dda0696f5d741)

After running for 5 minutes, it has only imported 2 MiB of the PCAP file.

```console
alois@laptop:~/Documents/edgesec ((12a7aa44...))$ timeout 5m bash -c 'pv ~/Downloads/2021_11_24_Active.pcap | ./build/linux/src/recap -f /tmp/recap-pcap-benchmark.sqlite'
Opened db at /tmp/recap-pcap-benchmark.sqlite
1.98MiB 0:04:59 [12.0KiB/s] [>                              ]  0% ETA 9:19:41:17
```

#### After (commit ce3c5f5b93f27ac864997f8f8cdc243e0caa23c8)

After running for 2:18 minutes, it had imported 5.48 GiB of the PCAP file (i.e. all of it).

Interestingly, it looks like it's now bottlenecked by CPU speed.

```console
alois@laptop:~/Documents/edgesec (optimize-recap)$ timeout 5m bash -c 'pv ~/Downloads/2021_11_24_Active.pcap | ./build/linux/src/recap -f /tmp/recap-pcap-benchmark.sqlite'
Opened db at /tmp/recap-pcap-benchmark.sqlite
5.48GiB 0:02:18 [40.4MiB/s] [================================>] 100%            
Processed pcap size = 5883456163 bytes
Processed packets = 18555332
```

I also tried compiling with `-DCMAKE_BUILD_TYPE=Release` (e.g. build with optimizations) and it actually seemed to slow things down to 2:40 minutes, but maybe that's just because my laptop has a terrible quality heatsink and it was fully saturated :laughing:

### Notes

I've also done a bit of minor refactoring while I was in the function, so please review this PR commit by commit!

It's up to you if this closes https://github.com/nqminds/edgesec/pull/347.
The optimizations in that PR should be cumulative with this PR, but it is at the cost of adding additional danger.

mereacre reviewed (commented):
Very good. There's one option when immediate transaction are not desirable, i.e., the case of continuous data ingestion from the libpcap stream. Currently I'm working on a branch that adds libpcap to recap. So ideally we need to add a flag "-u" (or similar) that allows user to use the commit functionality and this functionality is diasble when a continuous stream is used.

aloisklink commented:
> Very good. There's one option when immediate transaction are not desirable, i.e., the case of continuous data ingestion from the libpcap stream. Currently I'm working on a branch that adds libpcap to recap. So ideally we need to add a flag "-u" (or similar) that allows user to use the commit functionality and this functionality is diasble when a continuous stream is used.

Added in https://github.com/nqminds/edgesec/pull/348/commits/dad2b981e7b287e6d94fccbb75f9430b89d0e599!

This is what the new `recap -h` looks like:

```console
alois@laptop:~/Documents/edgesec (optimize-recap)$ ./build/linux/src/recap -h
recap app version 0.1.0-alpha.0
Usage:
	recap [-p filename] [-f filename] [-i interface] [-k] [-d] [-h] [-v]
		 [-t {SINGLE_TRANSACTION,DISABLED}]
Run capture on an input pcap file and output to a capture db or pipe.

Options:
	-p filename	 Path to the pcap file name
	-f filename	 Path to the capture db or pipe
	-i interface	 Interface name to save to db
	-t TRANS_TYPE	 Controls how SQLITE transactions are used.
	             	 Allowed values:
	             	   SINGLE_TRANSACTION: Use a single SQLITE transaction.
	             	   	 Provides maximum insertion performance.
	             	   DISABLED: Don't use SQLITE transactions (very slow).
	             	   (default) Pick automatically based on input.
	-k		 Pipe to file
	-d		 Verbosity level (use multiple -dd... to increase)
	-h		 Show help
	-v		 Show app version

Copyright NQMCyber Ltd
```

Basically, you can pass `-t SINGLE_TRANSACTION` to add everything in a super fast single transaction, `-t DISABLED` to be 3000x slower, or leave it blank, and `recap` will automatically choose an option for you (e.g. can be disabled for continuous streams).

I've also added some basic tests in https://github.com/nqminds/edgesec/commit/c4a245d1d90995c5495ba869ec70b7e1d5a93417 for the following cases in recap:

  - `recap -v`
  - `recap -f tmpdb -p pcap`
  - `recap -f tmpdb -p pcap -t SINGLE_TRANSACTION`
    (insert pcap into sqlite db in a single transaction)
  - `recap -f tmpdb -p pcap -t DISABLED`
    (insert pcap into sqlite db without transactions) (**warning, this takes like 5 seconds to complete, so it turns the original 2.5 second tests to 7.5 seconds**)
  - `recap -f tmpdb -p pcap -t invalid-arg-here`
    (checks for specific error message)
  - cleanup tmpdb when all tests are done

mereacre reviewed (approved):
Great
